app/controllers/schools/add_participants_controller.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+module Schools
+  class AddParticipantsController < ::Schools::BaseController
+    FORM_SESSION_KEY = :add_participant_form
+    FORM_PARAM_KEY = :schools_add_participant_form
+
+    skip_after_action :verify_authorized
+    before_action :set_school_cohort
+
+    helper_method :add_participant_form
+
+    def start
+      session.delete(FORM_SESSION_KEY)
+      redirect_to action: :show, step: :type
+    end
+
+    def show
+      render current_step
+    end
+
+    def update
+      if add_participant_form.valid?(current_step)
+        add_participant_form.record_completed_step current_step
+        store_form_in_session
+        redirect_to action: :show, step: step_param(add_participant_form.next_step(current_step))
+      else
+        render current_step
+      end
+    end
+
+  private
+
+    def add_participant_form
+      return @add_participant_form if defined?(@add_participant_form)
+
+      @add_participant_form = AddParticipantForm.new(session[FORM_SESSION_KEY])
+      @add_participant_form.assign_attributes(add_participant_form_params) if params[FORM_PARAM_KEY]
+
+      @add_participant_form
+    end
+
+    def store_form_in_session
+      session[FORM_SESSION_KEY] = add_participant_form.attributes
+    end
+
+    def current_step
+      params[:step].tr("-", "_").to_sym
+    end
+
+    def step_param(step)
+      step.to_s.tr("_", "-")
+    end
```

Comment:
You won't be shocked when I say I don't particularly like this. There are gems that perhaps do a better job at it, but more fundamentally, I don't like to have a single `update` for many steps. If a controller is meant to handle three update pages, maybe it should have three similar actions.

Reply:
This is again a result of Rails convention of resource-oriented thinking messing with a non-resourcefull logic. There is no resourcefull "update" here, we simply populate the form/domai operation object. No domain logic is being executed until the last step (that will have a separate action). Defining multiple steps would be really cumbersome.